Code/GraphMol/FileParsers/PDBParser.cpp

```diff
@@ -728,4 +733,219 @@ RWMol *PDBFileToMol(const std::string &fileName, bool sanitize, bool removeHs,
+      }
+      if (charge != 0) {
+        a->setFormalCharge(charge);
+      }
+    }
+
+    std::string tmp;
+
+    tmp = t.has_column(ATOMNAME) ? row[ATOMNAME] : "UNL";
+
+    AtomPDBResidueInfo* info = new AtomPDBResidueInfo(tmp);
+    a->setMonomerInfo(info);
+    if (row[GROUP_PDB][0] == 'H') {  // HETATM
+      info->setIsHeteroAtom(true);
+    }
+
+    int resno = 1;
+    if (t.has_column(RESNUM)) {
+      try {
+        resno = FileParserUtils::toInt(row[RESNUM]);
+      } catch (boost::bad_lexical_cast &) {
+        std::ostringstream errout;
+        errout << "Problem with residue number for PDB atom #" << a->getIdx();
+        throw FileParseException(errout.str());
+      }
+    }
+    info->setResidueNumber(resno);
+
+    tmp = t.has_column(RESNAME) ? row[RESNAME] : "UNL";
+    info->setResidueName(tmp);
+
+    tmp = t.has_column(ALT_ID) ? row[ALT_ID] : " ";
+    if (tmp == ".") { // "." used as non value instead of " "
+      tmp = " ";
+    }
+    info->setAltLoc(tmp);
+
+    tmp = t.has_column(CHAINID) ? row[CHAINID] : " ";
+    if (tmp == ".") {
+      tmp = " ";
+    }
+    info->setChainId(tmp);
+
+    tmp = t.has_column(INSCODE) ? row[INSCODE] : " ";
+    if (tmp == ".") {
+      tmp = " ";
+    }
+    info->setInsertionCode(tmp);
+
+    double occup = 1.0;
+    if (t.has_column(OCCUPANCY)) {
+      try {
+        occup = FileParserUtils::toDouble(row[OCCUPANCY]);
+      } catch (boost::bad_lexical_cast &) {
+        std::ostringstream errout;
+        errout << "Problem with occupancy for PDB atom #" << a->getIdx();
+        throw FileParseException(errout.str());
+      }
+    }
+    info->setOccupancy(occup);
+
+    double bfactor = 0.0;
+    if (t.has_column(TEMPFACTOR)) {
+      try {
+        bfactor = FileParserUtils::toDouble(row[TEMPFACTOR]);
+      } catch (boost::bad_lexical_cast &) {
+        std::ostringstream errout;
+        errout << "Problem with temperature factor for PDB atom #" << a->getIdx();
+        throw FileParseException(errout.str());
+      }
+    }
+    info->setTempFactor(bfactor);
+
+    // check all position columns were present
+    if (!t.has_column(CARTNX) || !t.has_column(CARTNY) || !t.has_column(CARTNZ)) {
+      continue;
+    }
+    RDGeom::Point3D pos;
+    try {
+      pos.x = FileParserUtils::toDouble(row[CARTNX]);
+      pos.y = FileParserUtils::toDouble(row[CARTNY]);
+      pos.z = FileParserUtils::toDouble(row[CARTNZ]);
+    } catch (boost::bad_lexical_cast &) {
+      std::ostringstream errout;
+      errout << "Problem with coordinates for PDB atom #" << a->getIdx();
+      throw FileParseException(errout.str());
+    }
+    Conformer *conf;
+    if (!mol->getNumConformers()) {
+      conf = new RDKit::Conformer(mol->getNumAtoms());
+      conf->set3D(pos.z != 0.0);  // worth checking for z coord?
+      mol->addConformer(conf, false);
+    } else {
+      conf = &mol->getConformer();
+      if (pos.z != 0.0) {
+        conf->set3D(true);
+      }
+    }
```

Comment:
I would simplify this whole thing by allocating the conformer and adding it to the molecule outside the loop over atoms.
And I think that for mmcif it's probably ok to skip the check for 3D and just assume that the molecule is 3D.